File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.java
class Solution {
    public int minFallingPathSum(int[][] grid) {
        
        int n = grid.length, firstMin = 0, secondMin = 0, position = -1;
        if(n == 1) return grid[0][0];
        
        for(int row = 0; row < n; ++row) {
            int firstMin2 = Integer.MAX_VALUE, secondMin2 = Integer.MAX_VALUE, position2 = -1;
            
            for(int col = 0; col < n; ++col) {
                int currentMin = col != position ? firstMin : secondMin;
                
                if(grid[row][col] + currentMin < firstMin2) {
                    secondMin2 = firstMin2;
                    firstMin2 = grid[row][col] + currentMin;
                    position2 = col;
                } else 
                    secondMin2 = Math.min(secondMin2, grid[row][col] + currentMin);
            }
            firstMin = firstMin2; 
            secondMin = secondMin2;
            position = position2;
        }
        return firstMin;
    }
}

File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.cpp
class Solution {
private:
    int findSum(int row, int col, const int &n, vector<vector<int>> &grid, vector<vector<int>> &dp) {
        if (col < 0 or col > n - 1) 
            return 1e9;
        if (row == 0)
            return grid[row][col];
        if (dp[row][col] != -1)
            return dp[row][col];
        int current = grid[row][col], ans = INT_MAX;
        for (int colTemp = 0; colTemp < n; ++colTemp) {
            if (colTemp == col) continue;
            ans = min(ans, findSum(row - 1, colTemp, n, grid, dp));
        }
        return dp[row][col] = current + ans;
    }
public:
    int minFallingPathSum(vector<vector<int>>& grid) {
        int n = size(grid), minSum = 1e9;
        vector<vector<int>> dp(n, vector<int> (n, -1));
        // try starting with each element in the last row
        for (int i = 0; i < n; ++i)
            minSum = min(minSum, findSum(n - 1, i, n, grid, dp));
        return minSum;
    }
};